packages/provider/tests/wallet.e2e.spec.ts

```diff
@@ -52,70 +52,202 @@ if (process.env.ONLY_E2E) {
+            gasLimit: 0,
             to: callReceiver.address,
             value: 0,
-            data: callReceiver.interface.functions.testCall.encode([5512, data2])
-          }
-        ]
+            data: callReceiver.interface.functions.testCall.encode([123, data])
+          }, {
+            from: wallet.address,
+            gasPrice: '20000000000',
+            gasLimit: 0,
+            to: callReceiver.address,
+            value: 0,
+            data: callReceiver.interface.functions.testCall.encode([123, '0x445566'])
+          }]
+
+          const arctxs = await toArcadeumTransactions(wallet, transactions)
+          const estimated = await relayer.estimateGasLimits(wallet.config, ARCADEUM_CONTEXT, ...arctxs)
+          expect((<any>estimated[0].gasLimit).toNumber()).to.be.above(600000)
+          expect((<any>estimated[0].gasLimit).toNumber()).to.be.below(1000000)
+          expect((<any>estimated[1].gasLimit).toNumber()).to.be.above(60000)
+          expect((<any>estimated[1].gasLimit).toNumber()).to.be.below(100000)
+        })
+        it('Should send meta-tx with estimate gasLimits', async () => {
+          await callReceiver.testCall(0, '0x')
+
+          const data = ethers.utils.randomBytes(1024)
+          const transactions = [{
+            from: wallet.address,
+            gasPrice: '20000000000',
+            gasLimit: 1000000,
+            to: callReceiver.address,
+            value: 0,
+            data: callReceiver.interface.functions.testCall.encode([743533, data])
+          }]
+
+          const arctxs = await toArcadeumTransactions(wallet, transactions, true)
+          const estimated = await relayer.estimateGasLimits(wallet.config, ARCADEUM_CONTEXT, ...arctxs)
+          const tx = await wallet.sendTransaction(estimated)
+          await tx.wait(2)
+
+          expect((await callReceiver.lastValA()).toString()).to.equal("743533")
+        })
+        it('Should send meta-txs with estimate gasLimits', async () => {
+          await callReceiver.testCall(0, '0x')
+
+          const data = ethers.utils.randomBytes(2048)
+          const transactions = [{
+            from: wallet.address,
+            gasPrice: '20000000000',
+            gasLimit: 0,
+            to: callReceiver.address,
+            value: 0,
+            data: callReceiver.interface.functions.testCall.encode([123, data])
+          }, {
+            from: wallet.address,
+            gasPrice: '20000000000',
+            gasLimit: 0,
+            to: callReceiver.address,
+            value: 0,
+            data: callReceiver.interface.functions.testCall.encode([7435, '0x445566'])
+          }]
 
-        const tx2 = await wallet.sendTransaction(transaction2)
-        await tx2.wait()
+          const arctxs = await toArcadeumTransactions(wallet, transactions, true)
+          const estimated = await relayer.estimateGasLimits(wallet.config, ARCADEUM_CONTEXT, ...arctxs)
+          const tx = await wallet.sendTransaction(estimated)
```

Comment:
the estimation stuff looks great, but where do we include the token fee from the client side..? that is, when a user is presented to sign a txn, we should also include the txn for the fee token, based on the estimate, so the user will sign the batch, etc..? is this included or does the wallet code need to include the fee txn into the batch upon confirmation?

Reply:
I think so, the wallet should take the batch of transactions and pass them through `estimateGasLimits`, then those transactions can be sent to `gasRefundOptions` to get a list of all possible refund options